students/arun_nalla/session07/html_render.py

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env python3
+
+"""
+A class-based system for rendering html.
+
+By Arun Nalla 11-30-2018
+"""
+
+# This is the framework for the base class
+
+class Element(object):
+    tag = "html"
+    indent = '   '
+
+
+    def __init__(self, content=None, **kwargs):
+        self.contents = []
+        if content is not None:
+            self.contents = [content]
+        self.kwargs = kwargs
+
+
+    def append(self, new_content):
+        self.contents.append(new_content)
+
+
+    def attributes(self):
+        attribute = ''
+        for key, value in self.kwargs.items():
+            attribute += ' {}="{}"'.format(key, value)
+        return attribute
+
+
+    def _open_tag(self):
+        open_tag = '<{}{}>'.format(self.tag, self.attributes())
+        return open_tag
+
+
+    def _close_tag(self):
+        close_tag = '</{}>'.format(self.tag)
+        return close_tag
+
+
+    def render(self, out_file, cur_ind = ''):
+        out_file.write(cur_ind)
+        out_file.write (self._open_tag())
+        out_file.write('\n')
+        for content in self.contents:
+            try:
+                content.render(out_file, cur_ind+self.indent)
+            except AttributeError:
+                out_file.write(cur_ind+self.indent)
+                out_file.write(content)
+            out_file.write("\n")
+        out_file.write(cur_ind)
+        out_file.write(self._close_tag())
+
+
+class Html(Element):
+
+    def render(self, out_file):
+        out_file.write('<!DOCTYPE html>\n')
+        super().render(out_file)
+
+
+class Body(Element):
+    tag = 'body'
+
+class P(Element):
+    tag = 'p'
+
+class Head(Element):
+    tag = 'head'
+
+class OneLineTag(Element):
+    tag = 'title'
+
```

Comment:
Minor: You may want to override `__init__()` here, so that `None` for `content` or any non-string `content` is not accepted for `OneLineTag` instances.